### PR TITLE
router: fix interaction between custom request headers, weighted clusters, and cluster_header

### PR DESCRIPTION
**Pull request by paul-r-gall**
Fix interaction between custom request headers, weighted clusters, and cluster_header

Signed-off-by: Paul Gallagher <pgal@google.com>

Commit Message: Fix interaction between custom request headers, weighted clusters, and cluster_header
Additional Description: This fixes a bug where when weighted clusters have a cluster header and `request_header_to_add` or similar fields, the (request|response)_header_to_(add|remove) fields are ignored.
Risk Level: Medium
Testing: Unit + Integration
Docs Changes: None
Release Notes:
Platform Specific Features:

### CODE REVIEW

File: source/common/router/config_impl.cc [hunk 1]
```diff
@@ -1267,7 +1270,7 @@ RouteConstSharedPtr RouteEntryImplBase::pickWeightedCluster(const Http::HeaderMa
     if (selected_value >= begin && selected_value < end) {
       if (!cluster->clusterHeaderName().get().empty() &&
           !headers.get(cluster->clusterHeaderName()).empty()) {
-        return pickClusterViaClusterHeader(cluster->clusterHeaderName(), headers);
+        return pickClusterViaClusterHeader(cluster->clusterHeaderName(), headers, cluster.get());
```
Comment (tyxia) [inline source/common/router/config_impl.cc:1273]: static_cast `cluster.get()` to `RouteEntryAndRoute*`? absl::implicit_cast should be best option here but I guess it is not available in Envoy code base.

**Review by tyxia — APPROVED** [review]
Thanks for fixing this!
LGTM modulo some nits.

File: test/common/router/config_impl_test.cc [hunk 2]
```diff
@@ -6285,6 +6285,53 @@ TEST_F(RouteMatcherTest, TestWeightedClusterHeaderManipulation) {
...
+                host_rewrite_literal: "new_host1"
+  )EOF";
+
+  factory_context_.cluster_manager_.initializeClusters({"cluster1"}, {});
+  TestConfigImpl config(parseRouteConfigurationFromYaml(yaml), factory_context_, true);
+  NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info;
+
+  {
+    Http::TestRequestHeaderMapImpl headers = genHeaders("www.lyft.com", "/foo", "GET");
+    headers.addCopy("x-route-to-this-cluster", "cluster1");
```
[the remark this answers is not in the document]
Comment (paul-r-gall) [reply test/common/router/config_impl_test.cc:6319]: Did a similar thing to what you suggested; preserved `cluster1` as the cluster name, and changed the other added header values.

File: source/common/router/config_impl.h [hunk 3]
```diff
@@ -927,8 +926,10 @@ class RouteEntryImplBase : public RouteEntry,
                               ProtobufMessage::ValidationVisitor& validator,
                               absl::string_view current_route_name) const;
 
-  RouteConstSharedPtr pickClusterViaClusterHeader(const Http::LowerCaseString& cluster_header_name,
-                                                  const Http::HeaderMap& headers) const;
+  RouteConstSharedPtr
+  pickClusterViaClusterHeader(const Http::LowerCaseString& cluster_header_name,
+                              const Http::HeaderMap& headers,
+                              const RouteEntryAndRoute* route_selector_override = nullptr) const;
```
[the remark this answers is not in the document]
Comment (paul-r-gall) [reply source/common/router/config_impl.h]: Done.

**Review by tyxia — APPROVED** [review]

**Review by zuercher — DISMISSED** [review]
Thanks! Looks good. I think the windows test was a flake, so hopefully it'll pass on re-test.

**Review by adisuissa — COMMENTED** [review]
Thanks for fixing this.
LGTM.
Can you please add a small release note (under [bug_fixes](https://github.com/envoyproxy/envoy/blob/6e74a51a5a3c15202526c9cd275cf3ce566f858a/changelogs/current.yaml#L81))

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
Should be merged once CI is green.

**Pull request merged**